rasa/utils/tensorflow/layers.py

```diff
@@ -1085,3 +1140,314 @@ def call(
+            all_labels_embed, candidate_ids
+        )
+        candidate_labels_ids = self._get_candidate_values(all_labels_ids, candidate_ids)
+
+        # Determine how many distinct labels exist (highest label index)
+        max_label_id = tf.cast(tf.math.reduce_max(all_labels_ids), dtype=tf.int32)
+
+        # Convert the positive label ids to their one_hot representation.
+        # Note: -1 indices yield a zeros-only vector
+        batch_labels_one_hot = tf.one_hot(
+            tf.cast(tf.squeeze(batch_labels_ids, axis=-1), tf.int32),
+            max_label_id + 1,
+            axis=-1,
+        )  # (batch_size, max_num_labels_per_input, depth_needed)
+
+        # Collapse the extra dimension and convert to a multi-hot representation
+        # by aggregating all ones in the one-hot representation.
+        # Here tf.reduce_any is important and tf.reduce_sum
+        # cannot be used. Reason being that due to padding in label_ids,
+        # there can be more than one 1's at the 0'th index of the last dimension.
+        # Otherwise the loss function will be highly unstable.
+        batch_labels_multi_hot = tf.cast(
+            tf.math.reduce_any(tf.cast(batch_labels_one_hot, dtype=tf.bool), axis=-2),
+            tf.float32,
+        )  # (batch_size, depth_needed)
+
+        # Remove extra dimensions for gather
+        candidate_labels_ids = tf.squeeze(tf.squeeze(candidate_labels_ids, 1), -1)
+
+        # Collect the labels corresponding to negative
+        # label ids sampled for each data point of the batch earlier.
+        pos_neg_labels = tf.gather(
+            batch_labels_multi_hot,
+            tf.cast(candidate_labels_ids, tf.int32),
+            batch_dims=1,
+            name="gather_labels",
+        )
+
+        pos_neg_labels = tf.cast(pos_neg_labels, tf.float32)
+
+        return (
+            pos_inputs_embed,
+            pos_labels_embed,
+            candidate_labels_embed,
+            pos_neg_labels,
+        )
+
+    def _get_candidate_indices(
+        self, batch_size: tf.Tensor, total_candidates: tf.Tensor
+    ) -> tf.Tensor:
+        """Returns batch of random candidate indices
+        
+        Args:
+            batch_size: Size of zero'th dimension
+            total_candidates: Size of first dimension
+
+        Returns:
+            tf.int32 tensor of shape (target_size, total_candidates) with values 
+            in [0, self.num_neg).
+        """
+        return layers_utils.random_indices(batch_size, self.num_neg, total_candidates)
+
+    @staticmethod
+    def _get_candidate_values(
+        x: tf.Tensor,  # (batch_size, ...)
+        candidate_ids: tf.Tensor,  # (batch_size, num_candidates)
+    ) -> tf.Tensor:
+        """Gathers candidate values according to IDs.
+        
+        Args:
+            x: Any tensor with at least one dimension
+            candidate_ids: Indicator for which candidates to gather
+
+        Returns:
```

Comment:
This one too. Can't we reuse `_sample_idxs` function from `SingleLabelDotProductLoss` by moving it to the base class and reusing it? We might need to change which axis is expanded / squeeze to get a general implementation but the idea of both the functions is the same, i.e. - Given a tensor and some ids, gather the values at those ids from the tensor.